abstracted micrometer feature so we can provide provider specific

JotBePunkt opened the pull request:
as wished by @dave08 in #1037 In also added an prometheus specific implementation that exposes the metrics endpoint with an optional authentication

cy6erGn0m commented on ktor-server/ktor-server-jetty/ktor-server-jetty-test-http2/build.gradle:
```diff
@@ -18,9 +18,9 @@ configure(tasks.jvmTest) {
-            def bootClasspath = configurations.boot.files
-            bootClasspath.collect { "-Xbootclasspath/p:${it.absolutePath}" }.iterator()
-        }
+       // jvmArgs {
```
?

JotBePunkt replied:
that should not have gone into the commit, there was an error preventing me from building.

JotBePunkt replied:
that should not have gone into the commit, there was an error preventing me from building. fixed with the next push

cy6erGn0m commented on ktor-features/ktor-metrics-micrometer/jvm/src/io/ktor/metrics/micrometer/MicrometerMetrics.kt:
```diff
@@ -42,177 +31,24 @@ import java.util.concurrent.atomic.*
 }
 
+/**
+ * the meter registry registered for this feature
```
Capitalize, same in other kdocs

cy6erGn0m commented on ktor-features/ktor-metrics-micrometer/jvm/src/io/ktor/metrics/micrometer/AbstractMicrometerMetrics.kt:
```diff
@@ -0,0 +1,212 @@
+        private val measureKey = AttributeKey<CallMeasure>("metrics")
+
+        private const val baseName: String = "ktor.http.server"
+        const val requestTimerName = "$baseName.requests"
```
kdoc or internal

cy6erGn0m commented on ktor-features/ktor-metrics-micrometer/jvm/src/io/ktor/metrics/micrometer/AbstractMicrometerMetrics.kt:
```diff
@@ -0,0 +1,212 @@
+
+            configuration.meterBinders.forEach { it.bindTo(configuration.registry) }
+
+            val phase = PipelinePhase("Metrics")
```
Please note that only one metrics feature could be installed since double phase insertion is not allowed. Phase need to be renamed or installed only if not yet present.

JotBePunkt replied:
I prefixed it with the actual class name, this should make it unique someone needed more than one Metric feature.

cy6erGn0m commented on ktor-features/ktor-metrics-micrometer/jvm/src/io/ktor/metrics/micrometer/AbstractMicrometerMetrics.kt:
```diff
@@ -0,0 +1,212 @@
+                }
+            }
+
+            val postSendPhase = PipelinePhase("MetricsPostSend")
```
Similar to `Metrics` phase

cy6erGn0m commented on ktor-features/ktor-metrics-micrometer-prometheus/jvm/src/io/ktor/metrics/micrometer/prometheus/installFeatureWithAuth.kt:
```diff
@@ -0,0 +1,32 @@
+    try {
+        feature(Authentication)
+    } catch (ex: MissingApplicationFeatureException) {
+        throw Exception(
```
Why `Exception`? I'd prefer to see some meaningful type

JotBePunkt replied:
fixed (I forgot how easy it is to declare a class in Kotlin...) In general this exception should not be caught during runtime, it shows the developer that they did not configure the feature correctly.

cy6erGn0m commented on ktor-features/ktor-metrics-micrometer-prometheus/jvm/src/io/ktor/metrics/micrometer/prometheus/installFeatureWithAuth.kt:
```diff
@@ -0,0 +1,32 @@
+import io.ktor.auth.*
+import io.ktor.routing.*
+
+fun Application.installMetricsEndpointWithAuth(config: MicrometerMetricsPrometheus.Configuration) {
```
We usually call such functions in declarative manner. For example, `routing { }` instead of `installRouting`, `authentication` instead of `installAuthentication` and so on. Also it makes sense to have micrometer in the name. Possible names would be `prometheusEndpoint` or `micrometerPrometheusEndpoint`. 

The other problem is that it requires `confi`g as a parameter while we are usually providing a lambda block to configure a feature

JotBePunkt replied:
made it internal and renamed it, its called when the feature is installed

cy6erGn0m commented on ktor-features/ktor-metrics-micrometer-prometheus/jvm/src/io/ktor/metrics/micrometer/prometheus/installFeatureWithAuth.kt:
```diff
@@ -0,0 +1,32 @@
+}
+
+private fun Application.checkIfAuthenticationFeatureIsInstalled() {
+    try {
```
According to your description, auth is optional however we actually always require auth.

JotBePunkt replied:
only if authentication() config is set in the configuration, if this is not set, we dont need the auth feature

dave08 replied:
Also, I would need, if possible, to be able to install the endpoint in a specific port (I have an internal port for prometheus scraping, and an external one that our proxy service exposes to the internet).

dave08 replied:
Maybe like providing an `var endpointPort: Int? = null`?

JotBePunkt replied:
hm... the next request would to specify the host name and now we start to reimplement the routing declaration. 
Alternative to the current implementation would be: The feature itself declares by default the endpoint under `/metrics` without any authentication, port selection etc. If you need some special routing feature you set a property `declareEndpoint = false` in the config, then you can declare the endpoint like you want it to be:
```
routing {
   port(1234) {
      authentication("superSecret") {
         //...
         get("/myMetricsPath") {
            prometheusMetrics()
          }
    }
}

cy6erGn0m replied:
The other solution is to provide routing entry for prometheus endpoint without any built-in auth. Like this:

```
fun Routing.prometheusEndpoint(...) { ... }

// so one can install it anywhere 
app.routing {
    route("/api") {
        prometheus { enable public metrics }
    }
    route("/admin") {
        authenticate(...) {
            // some admin actions go here
            prometheus { enable all metrics }
        }
    }
}
```

cy6erGn0m replied:
The advantage is that you don't need to care of auth at all. This is much more composable.

cy6erGn0m commented on ktor-features/ktor-metrics-micrometer-prometheus/jvm/src/io/ktor/metrics/micrometer/prometheus/MicrometerMetricsPrometheus.kt:
```diff
@@ -0,0 +1,89 @@
+    }
+
+private suspend fun PipelineContext<Unit, ApplicationCall>.prometheusMetrics() {
+    call.respondText(contentType = ContentType.parse(TextFormat.CONTENT_TYPE_004)) {
```
`CONTENT_TYPE_004` ?

JotBePunkt replied:
the value behind it is `text/plain; version=0.0.4; charset=utf-8`, I would take it as is so prometheus can easily parse it

JotBePunkt replied:
the value behind the constant is `text/plain; version=0.0.4; charset=utf-8` I would just use it like Prometheus needs it

cy6erGn0m replied:
Can we reassign it to a val-property with self-explaining name?

cy6erGn0m commented on ktor-features/ktor-metrics-micrometer-prometheus/jvm/src/io/ktor/metrics/micrometer/prometheus/MicrometerMetricsPrometheus.kt:
```diff
@@ -0,0 +1,89 @@
+        ): AbstractMicrometerMetrics<PrometheusMeterRegistry, Configuration> {
+            val feature = super.install(pipeline, configure)
+
+            if (feature.config.authenticationConfig.isEmpty()) {
```
This need to be unified to avoid duplication.

dave08 commented:
Thanks @JotBePunkt,two little questions though:

1. Why do we need to reimplement the feature? Wouldn't it be enough to have an extension function like in `ContentNegotiation`?
2. Can the metrics path be installed without authentication? My use case is that my Prometheus instance scraps our services inside the internal network between our VMs, so there isn't any real need for auth.

cy6erGn0m commented:
The main problem here is that there is huge abstraction, a lot of type parameters, constructor functions. Do we actually need it? Can't we replace inheritance with delegation/decoration somehow? So Prometheus endpoint feature could simply plug in into the micrometer feature

JotBePunkt commented:
>     1. Why do we need to reimplement the feature? Wouldn't it be enough to have an extension function like `gson { }` or `jackson { }` in `ContentNegotiation`?
The reusability is not really high, in contrast the json features, there ins only one thing that is encoded in that way (the meters registered in the prometheus meter registry) and there is more or less only one way to implement it so keep it simple here

> 2. Can the metrics path be installed without authentication? 

yes it can, just don't use the authenticate() method.
E.g.: 
https://github.com/ktorio/ktor/blob/159ad3707d9a95b932bdee275f9a2d56689fab45/ktor-features/ktor-metrics-micrometer-prometheus/jvm/test/io/ktor/metrics/micrometer/prometheus/MicrometerMetricsPrometheusTest.kt#L24

vs. (with auth)

https://github.com/ktorio/ktor/blob/159ad3707d9a95b932bdee275f9a2d56689fab45/ktor-features/ktor-metrics-micrometer-prometheus/jvm/test/io/ktor/metrics/micrometer/prometheus/MicrometerMetricsPrometheusTest.kt#L48-L50

dave08 commented:
But there isn't only Prometheus, there's also Dropwizard, and all the other implementations of the Micrometer facade (just like `gson { }` and `jackson { }`. If the feature's code is more or less the same for all those, it might be better to use a `ContentNegotiation`-type implementation... that was my original plan. Also I'd suppose that configs for the other types also vary...

Is there any real big changes in the base-Feature implementation that would make it simpler in the way you're currently doing it?

JotBePunkt commented:
> The main problem here is that there is huge abstraction, a lot of type parameters, constructor functions. Do we actually need it? Can't we replace inheritance with delegation/decoration somehow? So Prometheus endpoint feature could simply plug in into the micrometer feature

Yes, it looks complicate with all the generics and parameters. My main goals was to keep the usage a simple as possible. The alternatives would that the Prometheus feature only provides the metrics endpoint and
- the user needs to install the micrometer feature manually (which increases the complexity and the feature cannot provide prometheus specific default values for the config)
- the prometheus feature installs automatically the micrometer feature (which is not transparent for the user)

dave08 commented:
That might still be possible by having the Feature run a lambda that installs any endpoints the implementation requires, and have that provided by the `prometheus { }` config block...

JotBePunkt commented:
> But there isn't only Prometheus, there's also Dropwizard, and all the other implementations of the Micrometer facade (just like `gson { }` and `jackson { }`. If the feature's code is more or less the same for all those, it might be better to use a `ContentNegotiation`-type implementation... that was my original plan. Also I'd suppose that configs for the other types also vary...

As far as I know, Prometheus is the only metrics backend that needs a http endpoint in the application. The other implementation push their metrics to the metrics server or use an own endpoint (like JMX). 

The configs do vary but they are set in the registry instanciation e.g.

https://github.com/micrometer-metrics/micrometer/blob/127fa3265325cc894f368312ed8890b76a055d88/implementations/micrometer-registry-azure-monitor/src/main/java/io/micrometer/azuremonitor/AzureMonitorMeterRegistry.java#L55

https://github.com/micrometer-metrics/micrometer/blob/127fa3265325cc894f368312ed8890b76a055d88/implementations/micrometer-registry-atlas/src/main/java/io/micrometer/atlas/AtlasMeterRegistry.java#L53

In these cases it would be feasable to use the `MicrometerFeature` and provide a specificially configured `MeterRegistry`

> Is there any real big changes in the base-Feature implementation that would make it simpler in the way you're currently doing it?

JotBePunkt commented:
> That might still be possible by having the Feature run a lambda that installs any endpoints the implementation requires, and have that provided internally by the `prometheus { }` config block...

can you give an example how you would configure it?

cy6erGn0m commented:
(conflict is caused by metrics phase name clash fix in master)

JotBePunkt commented:
as discussed above, I removed the support for auth completely. I left default implementation of the metrics endpoint, which can be disabled so that the user can implement its own metricsEndpoints.
I switched the implementaiton to a pattern like the templating engines. In contrast to the Json features there is nothing to negotiate about the content type.

cy6erGn0m commented on ktor-features/ktor-metrics-micrometer-prometheus/jvm/src/io/ktor/metrics/micrometer/prometheus/MicrometerMetricsPrometheus.kt:
```diff
@@ -0,0 +1,149 @@
+         * Defines if a metrics endpoint is installed under "/metrics" for all metrics registered at [registry].
+         * Set to false if you want to declare a custom metric endpoint instead.
+         */
+        var installDefaultEndpoint = true
```
Instead of the flag something like `defaultEndpointRoute: String? = "/metrics"` can be used so both enable/disable and path customization will be accomplished at once.

cy6erGn0m commented on ktor-features/ktor-metrics-micrometer-prometheus/jvm/src/io/ktor/metrics/micrometer/prometheus/MicrometerMetricsPrometheus.kt:
```diff
@@ -0,0 +1,149 @@
+    fun filteredMetricFamilySamples(): Enumeration<Collector.MetricFamilySamples> =
+        object : Enumeration<Collector.MetricFamilySamples> {
+
+            private val internalCollection =
```
`registry.prometheusRegistry.metricFamilySamples().iterator().asSequence().filter(filter).iterator()` could replace a lot of the following code

cy6erGn0m commented on ktor-features/ktor-metrics-micrometer-prometheus/jvm/src/io/ktor/metrics/micrometer/prometheus/MicrometerMetricsPrometheus.kt:
```diff
@@ -0,0 +1,149 @@
+class MicrometerMetricsPrometheus(config: Configuration) :
+    AbstractMicrometerMetrics<PrometheusMeterRegistry, MicrometerMetricsPrometheus.Configuration>(config) {
+
+    internal fun process(value: MetricsContent) = object : OutgoingContent.WriteChannelContent() {
```
I'd prefer a separate internal class `PrometheusResponse`

cy6erGn0m commented on ktor-features/ktor-metrics-micrometer-prometheus/jvm/src/io/ktor/metrics/micrometer/prometheus/MicrometerMetricsPrometheus.kt:
```diff
@@ -0,0 +1,149 @@
+import kotlinx.coroutines.io.ByteWriteChannel
+import java.util.*
+
+class MicrometerMetricsPrometheus(config: Configuration) :
```
After removing unnecessary functionality, the redundant inheritance is even more obvious. I believe prometheus should be installed into the original micrometer metrics functionality. There is no need to inherit it. We need a separate Prometheus config and a routing function to install an endpoint. `meterRegistry` extension property could return it from the original feature as well.

JotBePunkt commented:
Sorry, I did not have much time to continue here. My nitial idea was to enable a specific configuration option for each of the MetricRegistry implementations. Prometheus is just one of them (but the one I am actively using), other implementation needs other config options. Therefore I used the implementation with the generics, which is not the favorite of @cy6erGn0m.

Oxymoron290 commented:
@JotBePunkt Understood, How can I help move this along. I would really like to see this feature as part of the project.

JotBePunkt commented:
Let me check, what I did so far. Afaik the Prometheus Implementation is kind of finished (I think we need to finish some details). What is missing are the other registry implementations for micrometer (or possibilities to configure them). I started to implement them, but its currenly in a state where it does not compile yet. What I could do is commit that to another branch and create a PR (as kind of preview, so you see what is planned).

JotBePunkt commented:
@Oxymoron290: Here is the continuation of the work: https://github.com/JotBePunkt/ktor/pull/2
I also asked the micrometer contributors if they like to see the configuration (Kotlin) classes in their repo, as other projects might want to profit from them too: 

https://github.com/micrometer-metrics/micrometer/issues/1607
